src/routes/documents/[id]-[slug]/annotate/+page.svelte

```diff
@@ -0,0 +1,192 @@
+<script lang="ts">
+  import type { Note, Sizes, ViewerMode } from "$lib/api/types";
+
+  import { afterNavigate, replaceState } from "$app/navigation";
+  import { page } from "$app/stores";
+
+  import { setContext } from "svelte";
+  import { writable, type Writable } from "svelte/store";
+  import { _ } from "svelte-i18n";
+
+  import ContentLayout from "$lib/components/layouts/ContentLayout.svelte";
+  import Paginator from "@/common/Paginator.svelte";
+  import PageToolbar from "$lib/components/common/PageToolbar.svelte";
+  import PDF from "$lib/components/documents/PDF.svelte";
+
+  import { pageHashUrl, pageFromHash } from "$lib/api/documents";
+  import { noteFromHash } from "$lib/api/notes";
+  import { scrollToPage } from "$lib/utils/scroll";
+
+  export let data;
+
+  // stores we need deeper in the component tree, available via context
+  const currentPage: Writable<number> = writable(1);
+  const activeNote: Writable<Note> = writable(null);
+  const mode: Writable<ViewerMode> = writable("document"); // only ever one mode on this route
+
+  setContext("currentPage", currentPage);
+  setContext("activeNote", activeNote);
+  setContext("mode", mode);
+
+  $: asset_url = data.asset_url;
+  $: document = data.document;
+  $: zoom = getDefaultZoom($mode);
+  $: zoomLevels = getZoomLevels($mode);
+
+  // lifecycle
+  afterNavigate(() => {
+    const { hash } = $page.url;
+
+    $currentPage = pageFromHash(hash);
+
+    if ($currentPage > 1) {
+      scrollToPage($currentPage);
+    }
+
+    const noteId = noteFromHash(hash);
+    if (noteId) {
+      $activeNote = document.notes.find((note) => note.id === noteId);
+    }
+  });
+
+  // pagination
+  function next() {
+    $currentPage = Math.min($currentPage + 1, document.page_count);
+    scrollToPage($currentPage);
+    replaceState(pageHashUrl($currentPage), {});
+  }
+
+  function previous() {
+    $currentPage = Math.max($currentPage - 1, 1);
+    scrollToPage($currentPage);
+    replaceState(pageHashUrl($currentPage), {});
+  }
+
+  function gotoPage(n: number) {
+    $currentPage = n;
+    scrollToPage($currentPage);
+    replaceState(pageHashUrl($currentPage), {});
+  }
+
+  function onHashChange(e: HashChangeEvent) {
+    const { hash } = new URL(e.newURL);
+    $currentPage = pageFromHash(hash);
+    scrollToPage($currentPage);
+
+    const noteId = noteFromHash(hash);
+    if (noteId) {
+      $activeNote = document.notes.find((note) => note.id === noteId);
+    }
+  }
```

Comment:
The `afterNavigate`, `next`, `previous`, `gotoPage`, and `onHashChange` functions on this page, the redaction page, and the parent page are all nearly identical—the only difference being line 42:

```
if ($currentPage > 1) {
```

Which has an extra condition in `src/routes/documents/[id]-[slug]/+page.svelte`:

```
if ($currentPage > 1 && ["document", "text"].includes($mode)) {
```

For consistency of behavior between viewer routes, now would be a good time to define a function for this common/duplicated functionality.

Reply:
I need to check if the stores work the same way in components and JS/TS files. That's why I didn't move the functions that use translations.